flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java

```diff
@@ -236,15 +228,30 @@ public void run() {
 					// Adjust number of records to fetch from the shard depending on current average record size
 					// to optimize 2 Mb / sec read limits
 					if (useAdaptiveReads) {
-						if (averageRecordSizeBytes != 0 && fetchIntervalMillis != 0) {
-							maxNumberOfRecordsPerFetch = (int) (KINESIS_SHARD_BYTES_PER_SECOND_LIMIT / (averageRecordSizeBytes * 1000L / fetchIntervalMillis));
+						if (averageRecordSizeBytes != 0 && runLoopTimeSeconds != 0) {
+							maxNumberOfRecordsPerFetch = (int) (KINESIS_SHARD_BYTES_PER_SECOND_LIMIT / (averageRecordSizeBytes / runLoopTimeSeconds));
 
 							// Ensure the value is not more than 10000L
 							maxNumberOfRecordsPerFetch = maxNumberOfRecordsPerFetch <= ConsumerConfigConstants.DEFAULT_SHARD_GETRECORDS_MAX ?
 									maxNumberOfRecordsPerFetch : ConsumerConfigConstants.DEFAULT_SHARD_GETRECORDS_MAX;
 						}
 					}
 
+					if (fetchIntervalMillis != 0) {
+							long preSleepTimeNanos = System.nanoTime();
+							long elapsedTimeNanos = preSleepTimeNanos - lastTimeNanos;
+							long sleepTimeMillis = fetchIntervalMillis - (elapsedTimeNanos / 1_000_000);
+
+							if (sleepTimeMillis > 0) {
+								Thread.sleep(sleepTimeMillis);
+							}
+							long postSleepTimeNanos = System.nanoTime();
+
+							runLoopTimeSeconds = (elapsedTimeNanos + (postSleepTimeNanos - preSleepTimeNanos)) / 1_000_000_000;
```

Comment:
You don't want to use whole seconds here.  This runLoopTimeSeconds variable should either be floating point or better yet just keep everything in nanoseconds throughout all of this code -- converting to other units only where needed.